File: src/main/java/me/syari/ss/discord/api/ThreadLocalReason.java
package me.syari.ss.discord.api;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class ThreadLocalReason {
    private static ThreadLocal<String> currentReason;

    private ThreadLocalReason() {
        throw new UnsupportedOperationException();
    }

    public static void setCurrent(@Nullable String reason) {
        if (reason != null) {
            if (currentReason == null)
                currentReason = new ThreadLocal<>();
            currentReason.set(reason);
        } else if (currentReason != null) {
            currentReason.remove();
        }
    }

    @Nullable
    public static String getCurrent() {
        return currentReason == null ? null : currentReason.get();
    }

    @NotNull
    public static Closable closable(@Nullable String reason) {
        return new Closable(reason);
    }

    public static class Closable implements AutoCloseable {
        private final String previous;

        public Closable(@Nullable String reason) {
            this.previous = getCurrent();
            setCurrent(reason);
        }

        @Override
        public void close() {
            setCurrent(previous);
        }
    }
}

File: src/main/kotlin/me/syari/ss/discord/requests/ThreadLocalReason.kt
package me.syari.ss.discord.requests

internal class ThreadLocalReason private constructor() {
    class Closable(reason: String?): AutoCloseable {
        private val previous = current

        override fun close() {
            current = previous
        }

        init {
            current = reason
        }
    }

    companion object {
        private var currentReason: ThreadLocal<String>? = null

        var current: String?
            get() = currentReason?.get()
            set(reason) {
                if (reason != null) {
                    if (currentReason == null) {
                        currentReason = ThreadLocal()
                    }
                    currentReason?.set(reason)
                } else if (currentReason != null) {
                    currentReason?.remove()
                }
            }

        fun closable(reason: String?): Closable {
            return Closable(reason)
        }
    }
}